src/test/java/org/javarosa/xpath/analysis/test/TreeReferenceAccumulatorTest.java

```diff
@@ -0,0 +1,157 @@
+
+
+        runAndTest(root,
+                "instance('expr')/ebase/element[@attr = current()/context_value]/value",
+                "instance('expr')/ebase/element/value",
+                "instance('expr')/ebase/element/@attr",
+                "instance('baseinstance')/base/element/context_value");
+
+        runAndTest(nestedRoot,
+                "instance('expr')/ebase/element[@attr = current()/context_value]/value",
+                "instance('expr')/ebase/element/value",
+                "instance('expr')/ebase/element/@attr",
+                "instance('nestedinstance')/nested/element/context_value");
+
+        runAndTest(root,
+                "../other_element",
+                "instance('baseinstance')/base/other_element");
+
+
+        runAndTest(root,
+                "instance('cars')/cars/car[../@lead_make = instance('models')/models/model[@year = current()/year]/make]/name",
+                "instance('cars')/cars/@lead_make",
+                "instance('cars')/cars/car/name",
+                "instance('models')/models/model/@year",
+                "instance('models')/models/model/make",
+                "instance('baseinstance')/base/element/year"
+                );
+
+
+
+
+    }
+
+
+
+    @Test
+    public void testInvalidReferences() throws XPathSyntaxException {
+        runForError("../relativeRef");
+        runForError("current()/relative");
+    }
+
+
+
+
+    private void runForError(String text) {
+        Set<TreeReference> references = runAndTest(text);
+        if (references != null) {
+            Assert.fail(String.format("Should have failed to analyse expression %s", text));
+        }
+    }
+
+
+    private Set<TreeReference> runAndTest(String text, String... matches) {
+        return runAndTest(null, text, matches);
+    }
+
+    /**
+     * Tests that running TreeReferenceAccumulatingAnalyzer on the XPath expression represented
+     * by @text results in accumulating all of the tree refs listed in @matches
+     */
+    private Set<TreeReference> runAndTest(EvaluationContext context, String text, String... matches)  {
+        XPathExpression expression;
+        try {
+            expression = XPathParseTool.parseXPath(text);
+        } catch(XPathSyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        TreeReferenceAccumulatingAnalyzer analyzer;
+        if (context == null) {
+            analyzer = new TreeReferenceAccumulatingAnalyzer();
+        } else {
+            analyzer = new TreeReferenceAccumulatingAnalyzer(context);
+        }
+
+        Set<TreeReference> results = analyzer.accumulate(expression);
+        if (results == null) {
+            return null;
+        }
+
+        Set<TreeReference> expressions = new HashSet<>();
+
+        for (String match : matches) {
+            expressions.add(XPathReference.getPathExpr(match).getReference());
+        }
+
+        Assert.assertEquals(text, expressions, results);
```

Comment:
Would be nice if the assertion message contained something more descriptive than just `text`

Reply:
The way these errors are output would be something like

```
Fail: TreeReferenceAccumulatorTest .testParsesWithContext
"instance('expr')/ebase/element[@attr = current()/context_value]/value"
expected:
                "instance('expr')/ebase/element/value",
                "instance('expr')/ebase/element/@attr",
                "instance('baseinstance')/base/element/context_value");
actual:
...
```
which seemed like the most helpful information that is available. junit also cuts off text aggressively, so I wanted to optimize for giving people as much context on what case was failing as possible (which is the XPath expression provided).

Reply:
got it, makes sense